File: brain_teasers/rotate_array.189.cpp
// In place reversal
// O(n) time
// O(1) space
class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        int n = nums.size();
        k %= n;
        std::reverse(nums.begin(), nums.end());
        std::reverse(nums.begin(), nums.begin() + k);
        std::reverse(nums.begin() + k, nums.end());
    }
};

// Manual reversal
// O(n) time
// O(1) space
class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        int n = nums.size();
        k %= n;
        reverseHelper(nums, 0, n - 1);
        reverseHelper(nums, 0, k - 1);
        reverseHelper(nums, k, n - 1);
    }
private:
    void reverseHelper(std::vector<int>& nums, int left, int right) {
        while (left < right) {
            // std::swap(nums[left], nums[right]);
            int temp = nums[left];
            nums[left] = nums[right];
            nums[right] = temp;
            left++;
            right--;
        }
    }
};

File: rotate_array.189.py
class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        k = k % len(nums) # important!!!
        n = len(nums)
        self.reverse_helper(nums, 0, n - 1)
        # or: nums[:] = nums[::-1]
        # or: nums.reverse()
        self.reverse_helper(nums, 0, k - 1)
        self.reverse_helper(nums, k, n - 1)

    def reverse_helper(self, nums, left, right) -> None:
        while left < right:
            nums[left], nums[right] = nums[right], nums[left]
            left += 1
            right -= 1
